crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -121,11 +354,143 @@ use std::{
+    fn apply(&mut self, system_meta: &SystemMeta, world: &mut World) {
+        self.apply(system_meta, world);
+    }
+
+    type Item<'world, 'state> = T::Item<'world, 'state>;
+
+    unsafe fn get_param<'world, 'state>(
+        state: &'state mut Self,
+        system_meta: &SystemMeta,
+        world: &'world World,
+        change_tick: u32,
+    ) -> Self::Item<'world, 'state> {
+        T::get_param(state, system_meta, world, change_tick).unwrap_or_else(|| {
+            panic!(
+                "Failed to get system param `{}`",
+                std::any::type_name::<T::Item<'world, 'state>>()
+            )
+        })
+    }
+}
+
+impl<T: SystemParam<Optional>> SystemParam<Infallible> for Option<T> {
+    type State = OptionState<T::State>;
+}
+
+#[doc(hidden)]
+pub struct OptionState<T: SystemParamState<Optional>>(T);
+
+// SAFETY: `SystemParamState<Optional>` must be safely implemented.
+unsafe impl<T: SystemParamState<Optional>> SystemParamState<Infallible> for OptionState<T> {
+    fn init(world: &mut World, system_meta: &mut SystemMeta) -> Self {
+        Self(T::init(world, system_meta))
+    }
+    #[inline]
+    fn new_archetype(&mut self, archetype: &Archetype, system_meta: &mut SystemMeta) {
+        self.0.new_archetype(archetype, system_meta);
+    }
+    #[inline]
+    fn apply(&mut self, system_meta: &SystemMeta, world: &mut World) {
+        self.0.apply(system_meta, world);
+    }
+
+    type Item<'world, 'state> = Option<T::Item<'world, 'state>>;
+
+    unsafe fn get_param<'world, 'state>(
+        state: &'state mut Self,
+        system_meta: &SystemMeta,
+        world: &'world World,
+        change_tick: u32,
+    ) -> Self::Item<'world, 'state> {
+        T::get_param(&mut state.0, system_meta, world, change_tick)
+    }
+}
+
+impl<T: SystemParam<Resultful>> SystemParam<Optional> for T {
+    type State = T::State;
+}
+
+// SAFETY: `SystemParamState<Resultful>` must be safely implemented.
+unsafe impl<T: SystemParamState<Resultful>> SystemParamState<Optional> for T {
+    fn init(world: &mut World, system_meta: &mut SystemMeta) -> Self {
+        T::init(world, system_meta)
+    }
+    #[inline]
+    fn new_archetype(&mut self, archetype: &Archetype, system_meta: &mut SystemMeta) {
+        self.new_archetype(archetype, system_meta);
+    }
+    #[inline]
+    fn apply(&mut self, system_meta: &SystemMeta, world: &mut World) {
+        self.apply(system_meta, world);
+    }
+
+    type Item<'world, 'state> = T::Item<'world, 'state>;
+
+    unsafe fn get_param<'world, 'state>(
+        state: &'state mut Self,
+        system_meta: &SystemMeta,
+        world: &'world World,
+        change_tick: u32,
+    ) -> Option<Self::Item<'world, 'state>> {
+        T::get_param(state, system_meta, world, change_tick).ok()
+    }
 }
 
-pub type SystemParamItem<'w, 's, P> = <<P as SystemParam>::State as SystemParamState>::Item<'w, 's>;
+impl<T: SystemParam<Resultful>> SystemParam<Infallible> for Result<T, Box<dyn Error>> {
```

Comment:
What's the use-case for a `Result<T, Box<dyn Error>>` SystemParam? I imagine there's not much you can do with the error except for just unwrapping it, and at that point you may as well just use `Option<T>` or `T` directly.

Reply:
This can be resolved: the &dyn Error is way more useful.